packages/browser-wallet/src/popup/store/token.ts

```diff
@@ -0,0 +1,117 @@
+                    };
+                })
+            )
+        ),
+    };
+});
+
+const accountTokensFamily = atomFamily<string, Atom<AsyncWrapper<Record<string, TokenIdAndMetadata[]>>>>(
+    (accountAddress) =>
+        atom((get) => {
+            const tokens = get(tokensAtom);
+            if (tokens.loading) {
+                return { loading: true, value: {} };
+            }
+            return { loading: false, value: tokens.value[accountAddress] };
+        })
+);
+
+export const currentAccountTokensAtom = atom<
+    AsyncWrapper<Record<string, TokenIdAndMetadata[]>>,
+    { contractIndex: string; newTokens: TokenIdAndMetadata[] },
+    Promise<void>
+>(
+    (get) => {
+        const currentAccount = get(selectedAccountAtom);
+        if (!currentAccount) {
+            return { loading: true, value: {} };
+        }
+        return get(accountTokensFamily(currentAccount));
+    },
+    async (get, set, { contractIndex, newTokens }) => {
+        const tokens = get(storedTokensAtom);
+        const tokenMetadata = get(tokenMetadataAtom);
+        if (tokens.loading || tokenMetadata.loading) {
+            throw new Error('Unable to update tokens while they are loading');
+        }
+        const currentAccount = get(selectedAccountAtom);
+        if (!currentAccount) {
+            throw new Error('Unable to update tokens for an account if there is no chosen account');
+        }
+        const accountCollections: Record<string, TokenStorage[]> = tokens.value[currentAccount] || {};
+        accountCollections[contractIndex] = newTokens.map((token) => ({
+            id: token.id,
+            metadataLink: token.metadataLink,
+        }));
+        const updatedTokens = { ...tokens.value };
+        updatedTokens[currentAccount] = accountCollections;
+        const newMetadata = tokenMetadata.value;
+        newTokens.forEach((token) => {
+            newMetadata[token.metadataLink] = token.metadata;
+        });
+        await set(tokenMetadataAtom, newMetadata);
+        return set(storedTokensAtom, updatedTokens);
+    }
+);
+
+const cbf = atomFamily<string, Atom<Promise<ContractBalances>>>((identifier: string) => {
+    const parts = identifier.split('.');
+
+    if ((parts as (string | undefined)[]).includes(undefined)) {
+        throw new Error('Could not get token balance family due to invalid input');
+    }
+
+    const [accountAddress, contractIndex] = parts;
+
+    return atom<Promise<ContractBalances>>(async (get) => {
+        const client = get(jsonRpcClientAtom);
+        const tokens = get(accountTokensFamily(accountAddress));
+
+        const tokenIds = tokens.value[contractIndex]?.map((t) => t.id) ?? [];
+
+        if (tokenIds.length === 0) {
+            return {};
+        }
+
+        return getContractBalances(client, contractIndex, tokenIds, accountAddress);
+    });
+});
+
+export const contractBalancesFamily = (accountAddress: string, contractIndex: string) =>
```

Comment:
So the family acts as a cache, basically. I guess we will need a way to invalidate the cache, because currently I can only force a reload of my assets by leaving the account page and coming back.
(But I assume we want it to be updated after some time :thinking:)

Reply:
Yes, thought that would be fine for now to just update it when you navigate to the account page. An atom family is just a map of string to atom, so yes it basically acts like a cache. I don't know how it would make sense to update the values, as we don't have a good way of knowing if the balance has changed, except for doing contract invoke every x seconds, which is probably not a good idea since it uses NRG?

An alternative could be to update all cis-2 balances when the public CCD balance changes. Of course that would only update the balance if you're the sender of the cis-2 transfer.

Reply:
I'm not sure I understand your concern, that it uses NRG? (it doesn't cost anything for the user)

It would be great if we could "invalidate the cache" after some time (like 2 minutes). Maybe that can be done in a separate PR.

Reply:
Decided on just going with a 1 minute interval for updating, and scrapping the ccd balance effect.